libraries/persistent_store/src/file.rs

```diff
@@ -0,0 +1,186 @@
+
+use crate::{Storage, StorageIndex, StorageResult};
+use alloc::borrow::Cow;
+use core::cell::RefCell;
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Seek, SeekFrom, Write};
+use std::path::Path;
+
+/// Simulates a flash storage using a host-based file.
+///
+/// This is usable for emulating authenticator hardware on VM hypervisor's host OS
+pub struct FileStorage {
+    // Options of the storage
+    options: FileOptions,
+
+    /// File for persisting contents of the storage
+    /// Reading data from File requires mutable reference, as seeking and reading data
+    /// changes file's current position.
+    /// All operations on backing file internally always first seek to needed position,
+    /// so it's safe to borrow mutable reference to backing file for the time of operation.
+    backing_file_ref: RefCell<File>,
+}
+
+/// Options for file-backed storage
+pub struct FileOptions {
+    /// Size of a word in bytes.
+    pub word_size: usize,
+
+    /// Size of a page in bytes.
+    pub page_size: usize,
+
+    /// Number of pages in storage
+    pub num_pages: usize,
+}
+
+impl FileStorage {
+    pub fn new(path: &Path, options: FileOptions) -> StorageResult<FileStorage> {
+        let mut backing_file_ref = RefCell::new(
+            OpenOptions::new()
+                .read(true)
+                .write(true)
+                .create(true)
+                .open(path)?,
+        );
+        let backing_file = backing_file_ref.get_mut();
+        let file_len = backing_file.metadata()?.len();
+        let store_len: u64 = (options.page_size * options.num_pages) as u64;
+
+        if file_len == 0 {
+            backing_file.seek(SeekFrom::Start(0))?;
+            let buf = vec![0xffu8; options.page_size];
+            for _ in 0..options.num_pages {
+                backing_file.write(&buf)?;
+            }
+        } else if file_len != store_len {
+            // FileStorage buffer should be of fixed size, opening previously saved file
+            // from storage of different size is not supported
+            panic!("Invalid file size {}, should be {}", file_len, store_len);
+        }
+        Ok(FileStorage {
+            options,
+            backing_file_ref,
+        })
+    }
+}
+
+impl Storage for FileStorage {
+    fn word_size(&self) -> usize {
+        self.options.word_size
+    }
+
+    fn page_size(&self) -> usize {
+        self.options.page_size
+    }
+
+    fn num_pages(&self) -> usize {
+        self.options.num_pages
+    }
+
+    fn max_word_writes(&self) -> usize {
+        usize::MAX
+    }
+
+    fn max_page_erases(&self) -> usize {
+        usize::MAX
+    }
+
+    fn read_slice(&self, index: StorageIndex, length: usize) -> StorageResult<Cow<[u8]>> {
+        let mut backing_file = self.backing_file_ref.borrow_mut();
+        backing_file.seek(SeekFrom::Start(
+            (index.page * self.page_size() + index.byte) as u64,
```

Comment:
We should call index.range(length, self)?.start to test the slice is valid.

Reply:
Done